File: quick/quick_parallel.py
from parallel_utils import parallel_ranges
from concurrent.futures import ThreadPoolExecutor


def _partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quicksort_parallel(arr, low=0, high=None, max_workers=None):
    work = list(arr)
    if high is None:
        high = len(work) - 1
    if low >= high:
        return work

    segments = [(low, high)]
    while segments:
        ranges = parallel_ranges(len(segments), max_workers=max_workers)
        next_segments_by_range = [[] for _ in ranges]

        def partition_range(args):
            range_index, segment_range = args
            children = next_segments_by_range[range_index]
            for segment_index in segment_range:
                lo, hi = segments[segment_index]
                if lo < hi:
                    pivot_index = _partition(work, lo, hi)
                    if lo < pivot_index - 1:
                        children.append((lo, pivot_index - 1))
                    if pivot_index + 1 < hi:
                        children.append((pivot_index + 1, hi))

        if len(ranges) == 1:
            partition_range((0, ranges[0]))
        else:
            with ThreadPoolExecutor(max_workers=len(ranges)) as executor:
                list(executor.map(partition_range, enumerate(ranges)))

        segments = [segment for children in next_segments_by_range for segment in children]

    return work

File: quick/quick_parallel.jl
using Polyester

function _partition_segment!(segment::Vector)
    pivot = segment[end]
    pivot_index = 0
    for j in 1:(length(segment) - 1)
        if segment[j] < pivot
            pivot_index += 1
            segment[pivot_index], segment[j] = segment[j], segment[pivot_index]
        end
    end
    pivot_index += 1
    segment[pivot_index], segment[end] = segment[end], segment[pivot_index]
    return pivot_index
end

function quicksort_parallel(arr::Vector)
    work = copy(arr)
    n = length(work)
    n < 2 && return work

    segments = Tuple{Int, Int}[(1, n)]
    while !isempty(segments)
        active_segments = Tuple{Int, Int}[]
        partitioned = Vector{Vector{eltype(work)}}()
        for (low, high) in segments
            if low < high
                push!(active_segments, (low, high))
                push!(partitioned, work[low:high])
            end
        end

        isempty(active_segments) && break

        pivot_offsets = Vector{Int}(undef, length(active_segments))
        @batch for idx in eachindex(partitioned)
            pivot_offsets[idx] = _partition_segment!(partitioned[idx])
        end

        next_segments = Tuple{Int, Int}[]
        for idx in eachindex(active_segments)
            low, high = active_segments[idx]
            work[low:high] = partitioned[idx]

            pivot_index = low + pivot_offsets[idx] - 1
            if low < pivot_index - 1
                push!(next_segments, (low, pivot_index - 1))
            end
            if pivot_index + 1 < high
                push!(next_segments, (pivot_index + 1, high))
            end
        end
        segments = next_segments
    end

    return work
end
